File: 257_Binary_Tree_Paths.cpp
/*
Given a binary tree, return all root-to-leaf paths.

Note: A leaf is a node with no children.

Example:
Input:

   1
 /   \
2     3
 \
  5
Output: ["1->2->5", "1->3"]

Explanation: All root-to-leaf paths are: 1->2->5, 1->3
*/
#include <iostream>
#include <vector>

using namespace std;

class Solution {

    struct TreeNode {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode(int x) : val(x), left(NULL), right(NULL) {}
    };
    
    public:
        vector<string> binaryTreePaths(TreeNode* root) {
            if(!root) return {};
            vector<string> path;
            string p = "";
            btTraversing(root, p, path);
            return path;
        }

        void btTraversing(TreeNode* root, string p, vector<string> &path) {
            if(!root) return;
            p += to_string(root->val);
            if(!root->left && !root->right) {
                path.push_back(p);
                return;
            }
            if(root->left) btTraversing(root->left, p + "->" , path);
            if(root->right) btTraversing(root->right, p + "->", path);
        }
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/257_Binary_Tree_Paths.py
from typing import List

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if not root:
            return []
        stack, res = [(root, "")], []
        while stack:
            node, path = stack.pop()
            if not node.left and not node.right:
                res.append(path + str(node.val))
            if node.left:
                stack.append((node.left, path + str(node.val) + "->"))
            if node.right:
                stack.append((node.right, path + str(node.val) + "->"))
        return res

    def binaryTreePaths2(self, root: TreeNode) -> List[str]:
        if not root:
            return []
        res = []
        self.dfs(res, "", root)
        return res

    def dfs(self, res: List[str], path: str, root: TreeNode):
        if not root.left and not root.right:
            res.append(path + str(root.val))
        if root.left:
            self.dfs(res, path + str(root.val) + "->", root.left)
        if root.right:
            self.dfs(res, path + str(root.val) + "->", root.right)
